airflow/jobs.py

```diff
@@ -2168,147 +2168,149 @@ def _process_backfill_task_instances(self,
             # or leaf to root, as otherwise tasks might be
             # determined deadlocked while they are actually
             # waiting for their upstream to finish
+            @provide_session
```

Comment:
Does this change also mean we open and close a connection for each TI? Does that have any impact on performance?

Reply:
Good point.
Not necessarily so, it just means that each TI gets its own session, which may or may not mapped to new connection as SQLAlchemy/Airflow does connection pooling. In fact, we use ti.refresh_from_db() a lot, where a new session object is created per call (https://github.com/apache/airflow/blob/c50a85146373bafb0cbf86850f834d63bd4dede8/airflow/models/__init__.py#L932).

Reply:
In general I think we should let SQLAlchemy do the pooling and close the sessions that we don't use anymore, instead of keeping them open and passing them around all the time.

Reply:
Personally I would prefer to have a `create_session`, since we commit the result on the last line anyway. If we do this properly, we shouldn't have to do `refresh_from_db` so often.

Reply:
I think SQLAlchemy does the pooling but is un-opinionated about how sessions are managed. The following access pattern is recommended per https://docs.sqlalchemy.org/en/latest/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it, which is what Airflow follows: https://github.com/apache/airflow/blob/c50a85146373bafb0cbf86850f834d63bd4dede8/airflow/utils/db.py#L37.

Reply:
Any other concerns? @Fokko @ashb

Reply:
Thanks for the reference link, no further concerns from my side.